#### Data Mining
#### Week 3
#### Swapnil Chavan

In [7]:
import pandas as pd
from textblob import TextBlob

#### PART 1 Using the TextBlob Sentiment Analyzer

##### 1. Import the movie review data as a data frame and ensure that the data is loaded properly

In [8]:
df1 = pd.read_csv("C://Users//Public//BagOfWords//labeledTrainData.tsv", sep="\t")


PermissionError: [Errno 13] Permission denied: 'C:/Users/Public/BagOfWords/labeledTrainData.tsv'

In [7]:
df1.head()

NameError: name 'df1' is not defined

##### 2. How many of each positive and negative reviews are there?
##### 3. How many of each positive and negative reviews are there?
##### 4. Use TextBlob to classify each movie review as positive or negative. Assume that a polarity score greater than or equal to zero is a positive sentiment and less than 0 is a negative sentiment.


In [55]:
def sentiment_calc(text):
    try:
        if TextBlob(text).sentiment.polarity >= 0:
            i = 1
        else:
            i = 0
        
        return i
    except:
        return None


In [56]:
df1['sentiment'] = df1['review'].apply(sentiment_calc)

In [57]:
df1.head()

id                                             review  sentiment
0  12311_10  Naturally in a film who's main themes are of m...          1
1    8348_2  This movie is a disaster within a disaster fil...          1
2    5828_4  All in all, this is a movie for kids. We saw i...          1
3    7186_2  Afraid of the Dark left me with the impression...          0
4   12128_7  A very accurate depiction of small time mob li...          1

In [58]:
s = df1['sentiment'] 
s_pos = ( s > 0 )
s_neg = ( s == 0 )
s_pos.sum()


18935

In [59]:
s_neg.sum()

6065

##### we saw there are 18935 positive sentiments and 6065 negative sentiments

Let's see how the new column data looks like

In [60]:
df1.head(25)

id                                             review  sentiment
0   12311_10  Naturally in a film who's main themes are of m...          1
1     8348_2  This movie is a disaster within a disaster fil...          1
2     5828_4  All in all, this is a movie for kids. We saw i...          1
3     7186_2  Afraid of the Dark left me with the impression...          0
4    12128_7  A very accurate depiction of small time mob li...          1
5     2913_8  ...as valuable as King Tut's tomb! (OK, maybe ...          1
6     4396_1  This has to be one of the biggest misfires eve...          1
7      395_2  This is one of those movies I watched, and won...          1
8    10616_1  The worst movie i've seen in years (and i've s...          0
9     9074_9  Five medical students (Kevin Bacon, David Labr...          1
10    9252_3  'The Mill on the Floss' was one of the lesser ...          1
11    9896_9  I just saw this film at the phoenix film festi...          1
12     574_4  \The Love Letter\" is one of those movies that...          1
13   11182_8  Another fantastic offering from the Monkey Isl...          1
14   11656_4  This was included on the disk \Shorts: Volume ...          1
15    2322_4  I'm not really much of an Abbott & Costello fa...          1
16    8703_1  This movie was dreadful. Biblically very inacc...          1
17    7483_1  I don't think I've ever gave something a 1/10 ...          0
18   6007_10  Excellent story-telling and cinematography. Po...          1
19   12424_4  I completely forgot that I'd seen this within ...          1
20    4672_1  I like action movies. I have a softspot for \B...          0
21   10841_3  This is one of the worst Sandra Bullock movie ...          0
22    8954_7  Watched this flick on Saturday afternoon cable...          1
23    7392_1  I went to see \TKIA\" with high expectations, ...          1
24   10288_8  All credit to writer/director Gilles Mimouni w...          1

##### 5. Check the accuracy of this model. Is this model better than random guessing?

##### 5. use another prebuilt text sentiment analyzer, e.g., VADER, and repeat steps (3) and (4).

In [61]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

In [62]:
analyzer = SentimentIntensityAnalyzer()

In [ ]:
Define a method that returns positive/negative based on the compound parameter value.

In [64]:
df1['compound'] = [analyzer.polarity_scores(x)['compound'] for x in df1['review']]

In [65]:
df1['sentiment_vader'] = df1['compound'].apply(vader_sent)

In [66]:
df1.head(10)

id                                             review  sentiment  \
0  12311_10  Naturally in a film who's main themes are of m...          1   
1    8348_2  This movie is a disaster within a disaster fil...          1   
2    5828_4  All in all, this is a movie for kids. We saw i...          1   
3    7186_2  Afraid of the Dark left me with the impression...          0   
4   12128_7  A very accurate depiction of small time mob li...          1   
5    2913_8  ...as valuable as King Tut's tomb! (OK, maybe ...          1   
6    4396_1  This has to be one of the biggest misfires eve...          1   
7     395_2  This is one of those movies I watched, and won...          1   
8   10616_1  The worst movie i've seen in years (and i've s...          0   
9    9074_9  Five medical students (Kevin Bacon, David Labr...          1   

   compound  sentiment_vader  
0    0.9692                1  
1   -0.5303                0  
2    0.9675                1  
3    0.4767                1  
4    0.9564                1  
5    0.9900                1  
6    0.9080                1  
7   -0.9269                0  
8   -0.9545                0  
9   -0.8020                0

#### PART 2 Prepping Text for a Custom Model

##### 1. Convert all text to lowercase letters.

In [3]:
df1['review'] = df1['review'].str.lower()

##### 2. Remove punctuation and special characters from the text.

In [4]:
def punctuation_removal(r):
    punc = '!()-[]{};:""\,<>./?@#$%^&*_~'''
    for ele in r:
        if ele in punc:
            r = r.replace(ele, "")
    return r

In [5]:
df1['review']=df1['review'].apply(punctuation_removal)

In [6]:
from nltk.stem import PorterStemmer

In [10]:
ps = PorterStemmer()

In [11]:
df1['review'] = [ps.stem(x) for x in df1['review']]

In [12]:
df1.head()

id                                             review
0  12311_10  naturally in a film who's main themes are of m...
1    8348_2  this movie is a disaster within a disaster fil...
2    5828_4  all in all this is a movie for kids we saw it ...
3    7186_2  afraid of the dark left me with the impression...
4   12128_7  a very accurate depiction of small time mob li...

In [13]:
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

In [14]:
stop_words = set(stopwords.words('english'))
df1['review'] = df1['review'].apply(lambda x: [item for item in str(x).split() if item not in stop_words])

In [15]:
df1.head()

id                                             review
0  12311_10  [naturally, film, who's, main, themes, mortali...
1    8348_2  [movie, disaster, within, disaster, film, full...
2    5828_4  [movie, kids, saw, tonight, child, loved, one,...
3    7186_2  [afraid, dark, left, impression, several, diff...
4   12128_7  [accurate, depiction, small, time, mob, life, ...

In [17]:
from sklearn.feature_extraction.text import CountVectorizer

In [18]:
vectorizer = CountVectorizer()

In [22]:
df1['review1']= df1['review'].apply(lambda x: " ".join(x))

In [31]:
df1.head()

id                                             review  \
0  12311_10  [naturally, film, who's, main, themes, mortali...   
1    8348_2  [movie, disaster, within, disaster, film, full...   
2    5828_4  [movie, kids, saw, tonight, child, loved, one,...   
3    7186_2  [afraid, dark, left, impression, several, diff...   
4   12128_7  [accurate, depiction, small, time, mob, life, ...   

                                             review1  
0  naturally film who's main themes mortality nos...  
1  movie disaster within disaster film full great...  
2  movie kids saw tonight child loved one point k...  
3  afraid dark left impression several different ...  
4  accurate depiction small time mob life filmed ...

In [35]:
matrix = vectorizer.fit_transform(df1.review1)
#counts = pd.DataFrame(matrix.toarray(),
  #                index=df1.id,
  #                columns=vectorizer.get_feature_names())

In [37]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [38]:
tfidf = TfidfVectorizer()

In [39]:
tfidf_matrix = tfidf.fit_transform(df1.review1)